signer/src/dkg/verification.rs

```diff
@@ -0,0 +1,804 @@
+                WstsNetMessageType::NonceResponse
+            }
+            _ => WstsNetMessageType::SignatureShareResponse,
+        }
+    }
+
+    /// Enqueues a message to be processed by the [`FrostCoordinator`] when
+    /// in the correct state.
+    fn enqueue_message(&mut self, sender: PublicKey, msg: wsts::net::Message) -> Result<(), Error> {
+        let msg_type: WstsNetMessageType = (&msg).into();
+
+        let current_count = self.total_message_count(msg_type);
+        let limit = self.message_type_limit(msg_type);
+
+        // If we've already received the maximum number of messages of this
+        // type, we don't enqueue it and return an error. Note: This may be
+        // unnecessary after changing to deduplicate messages by sender, but at
+        // least helps to cap the number of messages that we can enqueue if
+        // somehow there's messages from unknown pubkeys being dumped here.
+        //
+        // Note that this check is also our implicit check to ensure that only
+        // the allowed message types for signing can be enqueued, as the
+        // remaining WSTS message types have a limit of 0.
+        if current_count >= limit {
+            return Err(Error::MessageLimitExceeded {
+                message_type: msg_type,
+                expected: limit,
+                actual: current_count + 1,
+            });
+        }
+
+        // Enqueue the message under its message type and sender. We use the
+        // sender's public key to deduplicate messages from the same sender and
+        // message type; `insert_entry` will overwrite an existing message from the same
+        // sender, which we shouldn't have within the same round.
+        self.wsts_messages
+            .entry((msg_type, sender))
+            .insert_entry(QueuedMessage::new(msg));
+
+        Ok(())
+    }
+
+    /// Processes all queued messages that can be processed given the current
+    /// state of the [`FrostCoordinator`].
+    fn process_queued_messages(&mut self) -> Result<(), Error> {
+        let message_type_to_process = self.current_processable_message_type();
+
+        // Gets references to all pending messages of the given type that are
+        // currently stored. The returned messages are in arbitrary order, but
+        // this doesn't matter here since we're only processing the messages
+        // that the coordinator state machine can handle given its current
+        // state.
+
+        // We want to filter out processed messages.
+        let messages = self
+            .wsts_messages
+            .iter_mut()
+            .filter_map(|((msg_type, sender), msg)| {
+                if *msg_type == message_type_to_process && !msg.processed {
+                    Some((sender, msg))
+                } else {
+                    None
+                }
+            });
+
+        // Process all of the messages that we determined could be processed.
+        for (sender, msg) in messages {
```

Comment:
It could happen that we change the state while we still have messages of a certain type to process (eg, signers can "reset" their state by sending again a message, as in `enqueue_message` we overwrite with non processed messages). It should be fine -- the state machine would just ignore the messages -- but for completeness can we check, at the end of the loop, if the state machine state is still the one we expect, and if not break?

Reply:
This was changed around quite a bit in [c371a97](https://github.com/stacks-network/sbtc/pull/1324/commits/c371a974ef39859a568921efd032ea9ac274abb3), maybe have a new look and see if you think this comment is still relevant?

Reply:
It is still relevant: the possible "issue" (it should be handled by the inner state machine already, it's just to avoid relying on that unverified assumption) is that you may change your state (and thus what messages to process) while iterating over those messages, so you may end up consuming messages for the wrong state.

Reply:
Ohh now I get what you mean. Yeah, that's completely possible, but as you mentioned these will be ignored by the WSTS coordinator as it will have progressed to the "next" state. I didn't want to build in too much actual state logic here and rather delegate it to frost, but we could track the frost state before/after processing and return early if changed.

Reply:
Yeah I was thinking of something like that, ie if `self.current_processable_message_type` changed just return  -- we are already making decisions based on the underlying state, and we would keep the invariant here that is kinda defined at the beginning of this function (we pick the messages based on the inner state).

Reply:
Changed it to what I suggested above, as otherwise there's immutable+mutable borrow issues with the loop. The solution is to `collect()` the `messages_to_process` into a `Vec<_>`, but that's much less efficient.

[1566c74](https://github.com/stacks-network/sbtc/pull/1324/commits/1566c7488ff649309bf1040085f57ea0f44d029a)